Race detected during `TestWalletClaimGas` test execution

macOS, go 1.20, firstly noticed after https://github.com/nspcc-dev/neo-go/actions/runs/4789234667/jobs/8516850530?pr=2988 run.
```
2023-04-24T17:36:40.6068980Z === RUN   TestWalletClaimGas
2023-04-24T17:36:40.6069700Z     logger.go:130: 2023-04-24T17:35:50.606Z	INFO	initial gas supply is not set or wrong, setting default value	{"InitialGASSupply": "52000000"}
2023-04-24T17:36:40.6070540Z     logger.go:130: 2023-04-24T17:35:50.606Z	INFO	P2PNotaryRequestPayloadPool size is not set or wrong, setting default value	{"P2PNotaryRequestPayloadPoolSize": 1000}
2023-04-24T17:36:40.6071180Z     logger.go:130: 2023-04-24T17:35:50.606Z	INFO	MaxBlockSize is not set or wrong, setting default value	{"MaxBlockSize": 262144}
2023-04-24T17:36:40.6071890Z     logger.go:130: 2023-04-24T17:35:50.607Z	INFO	MaxBlockSystemFee is not set or wrong, setting default value	{"MaxBlockSystemFee": 900000000000}
2023-04-24T17:36:40.6072620Z     logger.go:130: 2023-04-24T17:35:50.607Z	INFO	MaxTransactionsPerBlock is not set or wrong, using default value	{"MaxTransactionsPerBlock": 512}
2023-04-24T17:36:40.6073370Z     logger.go:130: 2023-04-24T17:35:50.607Z	INFO	MaxValidUntilBlockIncrement is not set or wrong, using default value	{"MaxValidUntilBlockIncrement": 864000}
2023-04-24T17:36:40.6074260Z     logger.go:130: 2023-04-24T17:35:50.609Z	INFO	no storage version found! creating genesis block
2023-04-24T17:36:40.6075050Z     logger.go:130: 2023-04-24T17:35:50.616Z	INFO	MaxWebSocketClients is not set or wrong, setting default value	{"MaxWebSocketClients": 64}
2023-04-24T17:36:40.6075590Z     logger.go:130: 2023-04-24T17:35:50.616Z	INFO	starting rpc-server	{"endpoint": "localhost:0"}
2023-04-24T17:36:40.6076180Z === RUN   TestWalletClaimGas/missing_wallet_path
2023-04-24T17:36:40.6076700Z wallet path is mandatory and should be passed using (--wallet, -w) flags or via wallet config using --wallet-config flag
2023-04-24T17:36:40.6076870Z === RUN   TestWalletClaimGas/missing_address
2023-04-24T17:36:40.6076990Z address was not provided
2023-04-24T17:36:40.6077150Z === RUN   TestWalletClaimGas/invalid_address
2023-04-24T17:36:40.6077520Z can't find account for the address: NKuyBkoGdZZSLyPbJEetheRhMjeznFZszf
2023-04-24T17:36:40.6077690Z === RUN   TestWalletClaimGas/missing_endpoint
2023-04-24T17:36:40.6078030Z no RPC endpoint specified, use option '--rpc-endpoint' or '-r'
2023-04-24T17:36:40.6078200Z === RUN   TestWalletClaimGas/insufficient_funds
2023-04-24T17:36:40.6078330Z === NAME  TestWalletClaimGas
2023-04-24T17:36:40.6078940Z     logger.go:130: 2023-04-24T17:35:50.626Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6079580Z     logger.go:130: 2023-04-24T17:35:50.627Z	DEBUG	processing rpc request	{"method": "getnativecontracts", "params": "[]"}
2023-04-24T17:36:40.6080150Z     logger.go:130: 2023-04-24T17:35:50.644Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6081380Z     logger.go:130: 2023-04-24T17:35:50.645Z	DEBUG	processing rpc request	{"method": "invokescript", "params": "[CxAMFNwk6EPkEi4hj1BzCuuZPYVDDVeBDBTcJOhD5BIuIY9QcwrrmT2FQw1XgRTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I5 ]"}
2023-04-24T17:36:40.6081980Z     logger.go:130: 2023-04-24T17:35:50.648Z	DEBUG	processing rpc request	{"method": "getblockcount", "params": "[]"}
2023-04-24T17:36:40.6084620Z     logger.go:130: 2023-04-24T17:35:50.649Z	DEBUG	processing rpc request	{"method": "calculatenetworkfee", "params": "[AE6iJGu0P5gAAAAAAAAAAAAAAAAAAwAAAAHcJOhD5BIuIY9QcwrrmT2FQw1XgQEAVwsQDBTcJOhD5BIuIY9QcwrrmT2FQw1XgQwU3CToQ+QSLiGPUHMK65k9hUMNV4EUwB8MCHRyYW5zZmVyDBT1Y+pAvCg9TQ4FxI6jBbPyoHNA70FifVtSOQEAKAwhApd1MhCKuk+UUSF0us+LTXWT080Ozb18sexcCtOXfVMmQVbnsyc=]"}
2023-04-24T17:36:40.6084720Z EOF
2023-04-24T17:36:40.6085290Z     logger.go:130: 2023-04-24T17:35:50.653Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6085910Z     logger.go:130: 2023-04-24T17:35:50.655Z	DEBUG	processing rpc request	{"method": "getnativecontracts", "params": "[]"}
2023-04-24T17:36:40.6086660Z     logger.go:130: 2023-04-24T17:35:50.671Z	DEBUG	processing rpc request	{"method": "getnep17balances", "params": "[56c989e76f9a2ca05bb5caa6c96f524d905accd8]"}
2023-04-24T17:36:40.6087420Z     logger.go:130: 2023-04-24T17:35:50.673Z	DEBUG	processing rpc request	{"method": "getnep17balances", "params": "[56c989e76f9a2ca05bb5caa6c96f524d905accd8]"}
2023-04-24T17:36:40.6088000Z     logger.go:130: 2023-04-24T17:35:50.675Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6090320Z     logger.go:130: 2023-04-24T17:35:50.676Z	DEBUG	processing rpc request	{"method": "invokescript", "params": "[CwHoAwwU3CToQ+QSLiGPUHMK65k9hUMNV4EMFNjMWpBNUm/Jpsq1W6Asmm/niclWFMAfDAh0cmFuc2ZlcgwU9WPqQLwoPU0OBcSOowWz8qBzQO9BYn1bUjkLAwDodkgXAAAADBTcJOhD5BIuIY9QcwrrmT2FQw1XgQwU2MxakE1Sb8mmyrVboCyab+eJyVYUwB8MCHRyYW5zZmVyDBTPduKL0AYsSkeO41VhARMZ88+k0kFifVtSOQ== ]"}
2023-04-24T17:36:40.6090450Z ==================
2023-04-24T17:36:40.6090560Z WARNING: DATA RACE
2023-04-24T17:36:40.6090710Z Read at 0x00c0003041f0 by goroutine 2043:
2023-04-24T17:36:40.6091100Z   github.com/nspcc-dev/neo-go/pkg/core/native.copyNeoCache()
2023-04-24T17:36:40.6091640Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_neo.go:130 +0x184
2023-04-24T17:36:40.6092360Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*NeoCache).Copy()
2023-04-24T17:36:40.6092990Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_neo.go:121 +0x9d
2023-04-24T17:36:40.6093350Z   github.com/nspcc-dev/neo-go/pkg/core/dao.(*Simple).getCache()
2023-04-24T17:36:40.6093830Z       /Users/runner/work/neo-go/neo-go/pkg/core/dao/dao.go:931 +0xf6
2023-04-24T17:36:40.6094660Z   github.com/nspcc-dev/neo-go/pkg/core/dao.(*Simple).GetRWCache()
2023-04-24T17:36:40.6095140Z       /Users/runner/work/neo-go/neo-go/pkg/core/dao/dao.go:913 +0xae
2023-04-24T17:36:40.6095550Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).ModifyAccountVotes()
2023-04-24T17:36:40.6096070Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_neo.go:887 +0x84
2023-04-24T17:36:40.6096450Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).increaseBalance()
2023-04-24T17:36:40.6096960Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_neo.go:479 +0x6dc
2023-04-24T17:36:40.6097350Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).increaseBalance-fm()
2023-04-24T17:36:40.6097540Z       <autogenerated>:1 +0x93
2023-04-24T17:36:40.6097990Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*nep17TokenNative).updateAccBalance()
2023-04-24T17:36:40.6098520Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_nep17.go:190 +0x3a8
2023-04-24T17:36:40.6098960Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*nep17TokenNative).TransferInternal()
2023-04-24T17:36:40.6099500Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_nep17.go:230 +0x6e4
2023-04-24T17:36:40.6099900Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*nep17TokenNative).Transfer()
2023-04-24T17:36:40.6100420Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_nep17.go:118 +0x233
2023-04-24T17:36:40.6100830Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*nep17TokenNative).Transfer-fm()
2023-04-24T17:36:40.6101010Z       <autogenerated>:1 +0x6d
2023-04-24T17:36:40.6101330Z   github.com/nspcc-dev/neo-go/pkg/core/native.Call()
2023-04-24T17:36:40.6101850Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/interop.go:63 +0x15e4
2023-04-24T17:36:40.6102250Z   github.com/nspcc-dev/neo-go/pkg/core/interop.(*Context).SyscallHandler()
2023-04-24T17:36:40.6102770Z       /Users/runner/work/neo-go/neo-go/pkg/core/interop/context.go:323 +0x2ef
2023-04-24T17:36:40.6103170Z   github.com/nspcc-dev/neo-go/pkg/core/interop.(*Context).SyscallHandler-fm()
2023-04-24T17:36:40.6103350Z       <autogenerated>:1 +0x4b
2023-04-24T17:36:40.6103660Z   github.com/nspcc-dev/neo-go/pkg/vm.(*VM).execute()
2023-04-24T17:36:40.6104120Z       /Users/runner/work/neo-go/neo-go/pkg/vm/vm.go:1473 +0x1370
2023-04-24T17:36:40.6104420Z   github.com/nspcc-dev/neo-go/pkg/vm.(*VM).step()
2023-04-24T17:36:40.6104870Z       /Users/runner/work/neo-go/neo-go/pkg/vm/vm.go:472 +0x16f
2023-04-24T17:36:40.6105170Z   github.com/nspcc-dev/neo-go/pkg/vm.(*VM).Run()
2023-04-24T17:36:40.6105610Z       /Users/runner/work/neo-go/neo-go/pkg/vm/vm.go:444 +0x28f
2023-04-24T17:36:40.6106020Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).runScriptInVM()
2023-04-24T17:36:40.6106550Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:2205 +0x105
2023-04-24T17:36:40.6106950Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).invokescript()
2023-04-24T17:36:40.6107470Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:2017 +0xee
2023-04-24T17:36:40.6107860Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleIn()
2023-04-24T17:36:40.6108370Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:602 +0x927
2023-04-24T17:36:40.6108760Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleRequest()
2023-04-24T17:36:40.6109280Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:528 +0x18b
2023-04-24T17:36:40.6109690Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleHTTPRequest()
2023-04-24T17:36:40.6110210Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:508 +0x12b5
2023-04-24T17:36:40.6111000Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleHTTPRequest-fm()
2023-04-24T17:36:40.6111240Z       <autogenerated>:1 +0x57
2023-04-24T17:36:40.6111420Z   net/http.HandlerFunc.ServeHTTP()
2023-04-24T17:36:40.6111820Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/net/http/server.go:2122 +0x4d
2023-04-24T17:36:40.6112390Z   net/http.serverHandler.ServeHTTP()
2023-04-24T17:36:40.6112800Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/net/http/server.go:2936 +0x682
2023-04-24T17:36:40.6112930Z   net/http.(*conn).serve()
2023-04-24T17:36:40.6113320Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/net/http/server.go:1995 +0xbd4
2023-04-24T17:36:40.6114500Z   net/http.(*Server).Serve.func3()
2023-04-24T17:36:40.6115470Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/net/http/server.go:3089 +0x58
2023-04-24T17:36:40.6115510Z 
2023-04-24T17:36:40.6115710Z Previous write at 0x00c0003041f0 by goroutine 2019:
2023-04-24T17:36:40.6116270Z   github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).ComputeNextBlockValidators()
2023-04-24T17:36:40.6116850Z       /Users/runner/work/neo-go/neo-go/pkg/core/native/native_neo.go:1060 +0x2c4
2023-04-24T17:36:40.6117240Z   github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).GetValidators()
2023-04-24T17:36:40.6117750Z       /Users/runner/work/neo-go/neo-go/pkg/core/blockchain.go:2705 +0x94
2023-04-24T17:36:40.6118150Z   github.com/nspcc-dev/neo-go/pkg/consensus.(*service).getValidators()
2023-04-24T17:36:40.6118660Z       /Users/runner/work/neo-go/neo-go/pkg/consensus/consensus.go:682 +0x89
2023-04-24T17:36:40.6119060Z   github.com/nspcc-dev/neo-go/pkg/consensus.(*service).getValidators-fm()
2023-04-24T17:36:40.6119240Z       <autogenerated>:1 +0x64
2023-04-24T17:36:40.6119540Z   github.com/nspcc-dev/dbft.(*Context).Fill()
2023-04-24T17:36:40.6120450Z       /Users/runner/go/pkg/mod/github.com/nspcc-dev/dbft@v0.0.0-20230404091436-0108387748cd/context.go:268 +0x525
2023-04-24T17:36:40.6120850Z   github.com/nspcc-dev/dbft.(*Context).makePrepareRequest()
2023-04-24T17:36:40.6121710Z       /Users/runner/go/pkg/mod/github.com/nspcc-dev/dbft@v0.0.0-20230404091436-0108387748cd/send.go:19 +0x30
2023-04-24T17:36:40.6122090Z   github.com/nspcc-dev/dbft.(*DBFT).sendPrepareRequest()
2023-04-24T17:36:40.6122720Z       /Users/runner/go/pkg/mod/github.com/nspcc-dev/dbft@v0.0.0-20230404091436-0108387748cd/send.go:31 +0x4c
2023-04-24T17:36:40.6123030Z   github.com/nspcc-dev/dbft.(*DBFT).start()
2023-04-24T17:36:40.6123660Z       /Users/runner/go/pkg/mod/github.com/nspcc-dev/dbft@v0.0.0-20230404091436-0108387748cd/dbft.go:291 +0x18b
2023-04-24T17:36:40.6123950Z   github.com/nspcc-dev/dbft.(*DBFT).Start()
2023-04-24T17:36:40.6124570Z       /Users/runner/go/pkg/mod/github.com/nspcc-dev/dbft@v0.0.0-20230404091436-0108387748cd/dbft.go:83 +0x9a
2023-04-24T17:36:40.6125120Z   github.com/nspcc-dev/neo-go/pkg/consensus.(*service).Start()
2023-04-24T17:36:40.6125700Z       /Users/runner/work/neo-go/neo-go/pkg/consensus/consensus.go:283 +0x1ee
2023-04-24T17:36:40.6126110Z   github.com/nspcc-dev/neo-go/pkg/network.(*Server).tryStartServices()
2023-04-24T17:36:40.6126610Z       /Users/runner/work/neo-go/neo-go/pkg/network/server.go:551 +0x221
2023-04-24T17:36:40.6126940Z   github.com/nspcc-dev/neo-go/pkg/network.(*Server).Start()
2023-04-24T17:36:40.6127430Z       /Users/runner/work/neo-go/neo-go/pkg/network/server.go:273 +0x298
2023-04-24T17:36:40.6127830Z   github.com/nspcc-dev/neo-go/internal/testcli.NewTestChain.func2()
2023-04-24T17:36:40.6128350Z       /Users/runner/work/neo-go/neo-go/internal/testcli/executor.go:167 +0x39
2023-04-24T17:36:40.6128360Z 
2023-04-24T17:36:40.6128500Z Goroutine 2043 (running) created at:
2023-04-24T17:36:40.6128650Z   net/http.(*Server).Serve()
2023-04-24T17:36:40.6129140Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/net/http/server.go:3089 +0x817
2023-04-24T17:36:40.6129590Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).Start.func1()
2023-04-24T17:36:40.6130570Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:363 +0x78
2023-04-24T17:36:40.6131070Z   github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).Start.func4()
2023-04-24T17:36:40.6131600Z       /Users/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:368 +0x47
2023-04-24T17:36:40.6131620Z 
2023-04-24T17:36:40.6131750Z Goroutine 2019 (running) created at:
2023-04-24T17:36:40.6132580Z   github.com/nspcc-dev/neo-go/internal/testcli.NewTestChain()
2023-04-24T17:36:40.6133320Z       /Users/runner/work/neo-go/neo-go/internal/testcli/executor.go:167 +0x1151
2023-04-24T17:36:40.6133790Z   github.com/nspcc-dev/neo-go/internal/testcli.NewExecutorWithConfig()
2023-04-24T17:36:40.6134320Z       /Users/runner/work/neo-go/neo-go/internal/testcli/executor.go:193 +0x4e4
2023-04-24T17:36:40.6134670Z   github.com/nspcc-dev/neo-go/internal/testcli.NewExecutor()
2023-04-24T17:36:40.6135180Z       /Users/runner/work/neo-go/neo-go/internal/testcli/executor.go:176 +0x54
2023-04-24T17:36:40.6135560Z   github.com/nspcc-dev/neo-go/cli/wallet_test.TestWalletClaimGas()
2023-04-24T17:36:40.6136620Z       /Users/runner/work/neo-go/neo-go/cli/wallet/wallet_test.go:537 +0x3e
2023-04-24T17:36:40.6136810Z   testing.tRunner()
2023-04-24T17:36:40.6137240Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1576 +0x216
2023-04-24T17:36:40.6137390Z   testing.(*T).Run.func1()
2023-04-24T17:36:40.6137800Z       /Users/runner/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1629 +0x47
2023-04-24T17:36:40.6137900Z ==================
2023-04-24T17:36:40.6138600Z     logger.go:130: 2023-04-24T17:35:50.682Z	DEBUG	processing rpc request	{"method": "getblockcount", "params": "[]"}
2023-04-24T17:36:40.6141500Z     logger.go:130: 2023-04-24T17:35:50.683Z	DEBUG	processing rpc request	{"method": "calculatenetworkfee", "params": "[AEz0555ofzABAAAAAAAAAAAAAAAAAwAAAAHYzFqQTVJvyabKtVugLJpv54nJVgEAuAsB6AMMFNwk6EPkEi4hj1BzCuuZPYVDDVeBDBTYzFqQTVJvyabKtVugLJpv54nJVhTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I5CwMA6HZIFwAAAAwU3CToQ+QSLiGPUHMK65k9hUMNV4EMFNjMWpBNUm/Jpsq1W6Asmm/niclWFMAfDAh0cmFuc2ZlcgwUz3bii9AGLEpHjuNVYQETGfPPpNJBYn1bUjkBACoRDCECs2Ir9AF73+MXxYrtX0x1PyBrfbiWBG+n13S7xL9/jcIRQZ7Q3Do=]"}
2023-04-24T17:36:40.6145230Z     logger.go:130: 2023-04-24T17:35:50.685Z	DEBUG	processing rpc request	{"method": "sendrawtransaction", "params": "[AEz0555ofzABAAAAANxBFAAAAAAAAwAAAAHYzFqQTVJvyabKtVugLJpv54nJVgEAuAsB6AMMFNwk6EPkEi4hj1BzCuuZPYVDDVeBDBTYzFqQTVJvyabKtVugLJpv54nJVhTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I5CwMA6HZIFwAAAAwU3CToQ+QSLiGPUHMK65k9hUMNV4EMFNjMWpBNUm/Jpsq1W6Asmm/niclWFMAfDAh0cmFuc2ZlcgwUz3bii9AGLEpHjuNVYQETGfPPpNJBYn1bUjkBQgxAy+wc0LnG4s4f3QqJmt2Gke7EH7Y2KAPJKuq/q+0IGfHQwKlvIIbIxDoAEsU0Mk9PoUI7H/aWhAgN4b+PHkFINyoRDCECs2Ir9AF73+MXxYrtX0x1PyBrfbiWBG+n13S7xL9/jcIRQZ7Q3Do=]"}
2023-04-24T17:36:40.6146080Z     logger.go:130: 2023-04-24T17:35:50.690Z	DEBUG	done processing headers	{"headerIndex": 1, "blockHeight": 0, "took": "524.22µs"}
2023-04-24T17:36:40.6146760Z     logger.go:130: 2023-04-24T17:35:50.793Z	DEBUG	done processing headers	{"headerIndex": 2, "blockHeight": 1, "took": "296.701µs"}
2023-04-24T17:36:40.6147420Z     logger.go:130: 2023-04-24T17:35:50.894Z	DEBUG	done processing headers	{"headerIndex": 3, "blockHeight": 2, "took": "324.058µs"}
2023-04-24T17:36:40.6148020Z     logger.go:130: 2023-04-24T17:35:50.895Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6148670Z     logger.go:130: 2023-04-24T17:35:50.896Z	DEBUG	processing rpc request	{"method": "getnativecontracts", "params": "[]"}
2023-04-24T17:36:40.6149250Z     logger.go:130: 2023-04-24T17:35:50.909Z	DEBUG	processing rpc request	{"method": "getversion", "params": "[]"}
2023-04-24T17:36:40.6150490Z     logger.go:130: 2023-04-24T17:35:50.910Z	DEBUG	processing rpc request	{"method": "invokescript", "params": "[CxAMFNwk6EPkEi4hj1BzCuuZPYVDDVeBDBTcJOhD5BIuIY9QcwrrmT2FQw1XgRTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I5 ]"}
2023-04-24T17:36:40.6151630Z     logger.go:130: 2023-04-24T17:35:50.912Z	DEBUG	processing rpc request	{"method": "getblockcount", "params": "[]"}
2023-04-24T17:36:40.6154390Z     logger.go:130: 2023-04-24T17:35:50.913Z	DEBUG	processing rpc request	{"method": "calculatenetworkfee", "params": "[AK1BcPu0P5gAAAAAAAAAAAAAAAAABgAAAAHcJOhD5BIuIY9QcwrrmT2FQw1XgQEAVwsQDBTcJOhD5BIuIY9QcwrrmT2FQw1XgQwU3CToQ+QSLiGPUHMK65k9hUMNV4EUwB8MCHRyYW5zZmVyDBT1Y+pAvCg9TQ4FxI6jBbPyoHNA70FifVtSOQEAKAwhApd1MhCKuk+UUSF0us+LTXWT080Ozb18sexcCtOXfVMmQVbnsyc=]"}
2023-04-24T17:36:40.6157840Z     logger.go:130: 2023-04-24T17:35:50.914Z	DEBUG	processing rpc request	{"method": "sendrawtransaction", "params": "[AK1BcPu0P5gAAAAAAOi+EgAAAAAABgAAAAHcJOhD5BIuIY9QcwrrmT2FQw1XgQEAVwsQDBTcJOhD5BIuIY9QcwrrmT2FQw1XgQwU3CToQ+QSLiGPUHMK65k9hUMNV4EUwB8MCHRyYW5zZmVyDBT1Y+pAvCg9TQ4FxI6jBbPyoHNA70FifVtSOQFCDEBe7T1AU4Vac0jjrd78deU06Rk24kUwlrVrwZxHDqG0j21TK/DLBTQ3VJhiVXbN+7xVx83lB9zZdTJzQkapth3VKAwhApd1MhCKuk+UUSF0us+LTXWT080Ozb18sexcCtOXfVMmQVbnsyc=]"}
2023-04-24T17:36:40.6158520Z     logger.go:130: 2023-04-24T17:35:50.997Z	DEBUG	done processing headers	{"headerIndex": 4, "blockHeight": 3, "took": "291.203µs"}
2023-04-24T17:36:40.6159810Z     logger.go:130: 2023-04-24T17:35:51.018Z	INFO	shutting down RPC server	{"endpoint": "127.0.0.1:51017"}
2023-04-24T17:36:40.6160560Z     logger.go:130: 2023-04-24T17:35:51.019Z	INFO	persisted to disk	{"blocks": 4, "keys": 183, "headerHeight": 4, "blockHeight": 4, "took": "232.134µs"}
2023-04-24T17:36:40.6161330Z     testing.go:1446: race detected during execution of test
2023-04-24T17:36:40.6161670Z --- FAIL: TestWalletClaimGas (0.42s)
2023-04-24T17:36:40.6162200Z     --- PASS: TestWalletClaimGas/missing_wallet_path (0.00s)
2023-04-24T17:36:40.6162600Z     --- PASS: TestWalletClaimGas/missing_address (0.00s)
2023-04-24T17:36:40.6162970Z     --- PASS: TestWalletClaimGas/invalid_address (0.00s)
2023-04-24T17:36:40.6163340Z     --- PASS: TestWalletClaimGas/missing_endpoint (0.00s)
2023-04-24T17:36:40.6163720Z     --- PASS: TestWalletClaimGas/insufficient_funds (0.03s)
```

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -48,7 +48,7 @@ type Ledger interface {
 	GetNextBlockValidators() ([]*keys.PublicKey, error)
 	GetStateRoot(height uint32) (*state.MPTRoot, error)
 	GetTransaction(util.Uint256) (*transaction.Transaction, uint32, error)
-	GetValidators() ([]*keys.PublicKey, error)
+	GetValidators() []*keys.PublicKey
 	PoolTx(t *transaction.Transaction, pools ...*mempool.Pool) error
 	SubscribeForBlocks(ch chan *coreb.Block)
 	UnsubscribeFromBlocks(ch chan *coreb.Block)
@@ -679,7 +679,7 @@ func (s *service) getValidators(txes ...block.Transaction) []crypto.PublicKey {
 	if txes == nil {
 		pKeys, err = s.Chain.GetNextBlockValidators()
 	} else {
-		pKeys, err = s.Chain.GetValidators()
+		pKeys = s.Chain.GetValidators()
 	}
 	if err != nil {
 		s.log.Error("error while trying to get validators", zap.Error(err))
@@ -725,7 +725,7 @@ func (s *service) newBlockFromContext(ctx *dbft.Context) block.Block {
 	var err error
 	cfg := s.Chain.GetConfig().ProtocolConfiguration
 	if cfg.ShouldUpdateCommitteeAt(ctx.BlockIndex) {
-		validators, err = s.Chain.GetValidators()
+		validators = s.Chain.GetValidators()
 	} else {
 		validators, err = s.Chain.GetNextBlockValidators()
 	}

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -677,9 +677,22 @@ func (s *service) getValidators(txes ...block.Transaction) []crypto.PublicKey {
 		err   error
 	)
 	if txes == nil {
+		// getValidators with empty args is used by dbft to fill the list of
+		// block's validators, thus should return validators from the current
+		// epoch without recalculation.
 		pKeys, err = s.Chain.GetNextBlockValidators()
 	} else {
-		pKeys = s.Chain.ComputeNextBlockValidators()
+		// getValidators with non-empty args is used by dbft to fill block's
+		// NextConsensus field, thus should return proper value for NextConsensus.
+		cfg := s.Chain.GetConfig().ProtocolConfiguration
+		if cfg.ShouldUpdateCommitteeAt(s.dbft.Context.BlockIndex) {
+			// Calculate NextConsensus based on the most fresh chain state.
+			pKeys = s.Chain.ComputeNextBlockValidators()
+		} else {
+			// Take the cached validators that are relevant for the current dBFT epoch
+			// to make NextConsensus.
+			pKeys, err = s.Chain.GetNextBlockValidators()
+		}
 	}
 	if err != nil {
 		s.log.Error("error while trying to get validators", zap.Error(err))
@@ -725,8 +738,11 @@ func (s *service) newBlockFromContext(ctx *dbft.Context) block.Block {
 	var err error
 	cfg := s.Chain.GetConfig().ProtocolConfiguration
 	if cfg.ShouldUpdateCommitteeAt(ctx.BlockIndex) {
+		// Calculate NextConsensus based on the most fresh chain state.
 		validators = s.Chain.ComputeNextBlockValidators()
 	} else {
+		// Take the cached validators that are relevant for the current dBFT epoch
+		// to make NextConsensus.
 		validators, err = s.Chain.GetNextBlockValidators()
 	}
 	if err != nil {

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -683,16 +683,9 @@ func (s *service) getValidators(txes ...block.Transaction) []crypto.PublicKey {
 		pKeys, err = s.Chain.GetNextBlockValidators()
 	} else {
 		// getValidators with non-empty args is used by dbft to fill block's
-		// NextConsensus field, thus should return proper value for NextConsensus.
-		cfg := s.Chain.GetConfig().ProtocolConfiguration
-		if cfg.ShouldUpdateCommitteeAt(s.dbft.Context.BlockIndex) {
-			// Calculate NextConsensus based on the most fresh chain state.
-			pKeys = s.Chain.ComputeNextBlockValidators()
-		} else {
-			// Take the cached validators that are relevant for the current dBFT epoch
-			// to make NextConsensus.
-			pKeys, err = s.Chain.GetNextBlockValidators()
-		}
+		// NextConsensus field, ComputeNextBlockValidators will return proper
+		// value for NextConsensus wrt dBFT epoch start/end.
+		pKeys = s.Chain.ComputeNextBlockValidators()
 	}
 	if err != nil {
 		s.log.Error("error while trying to get validators", zap.Error(err))
@@ -734,20 +727,7 @@ func (s *service) newBlockFromContext(ctx *dbft.Context) block.Block {
 		block.PrevStateRoot = sr.Root
 	}
 
-	var validators keys.PublicKeys
-	var err error
-	cfg := s.Chain.GetConfig().ProtocolConfiguration
-	if cfg.ShouldUpdateCommitteeAt(ctx.BlockIndex) {
-		// Calculate NextConsensus based on the most fresh chain state.
-		validators = s.Chain.ComputeNextBlockValidators()
-	} else {
-		// Take the cached validators that are relevant for the current dBFT epoch
-		// to make NextConsensus.
-		validators, err = s.Chain.GetNextBlockValidators()
-	}
-	if err != nil {
-		s.log.Fatal(fmt.Sprintf("failed to get validators: %s", err.Error()))
-	}
+	var validators = s.Chain.ComputeNextBlockValidators()
 	script, err := smartcontract.CreateDefaultMultiSigRedeemScript(validators)
 	if err != nil {
 		s.log.Fatal(fmt.Sprintf("failed to create multisignature script: %s", err.Error()))

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -48,7 +48,7 @@ type Ledger interface {
 	GetNextBlockValidators() ([]*keys.PublicKey, error)
 	GetStateRoot(height uint32) (*state.MPTRoot, error)
 	GetTransaction(util.Uint256) (*transaction.Transaction, uint32, error)
-	GetValidators() []*keys.PublicKey
+	ComputeNextBlockValidators() []*keys.PublicKey
 	PoolTx(t *transaction.Transaction, pools ...*mempool.Pool) error
 	SubscribeForBlocks(ch chan *coreb.Block)
 	UnsubscribeFromBlocks(ch chan *coreb.Block)
@@ -679,7 +679,7 @@ func (s *service) getValidators(txes ...block.Transaction) []crypto.PublicKey {
 	if txes == nil {
 		pKeys, err = s.Chain.GetNextBlockValidators()
 	} else {
-		pKeys = s.Chain.GetValidators()
+		pKeys = s.Chain.ComputeNextBlockValidators()
 	}
 	if err != nil {
 		s.log.Error("error while trying to get validators", zap.Error(err))
@@ -725,7 +725,7 @@ func (s *service) newBlockFromContext(ctx *dbft.Context) block.Block {
 	var err error
 	cfg := s.Chain.GetConfig().ProtocolConfiguration
 	if cfg.ShouldUpdateCommitteeAt(ctx.BlockIndex) {
-		validators = s.Chain.GetValidators()
+		validators = s.Chain.ComputeNextBlockValidators()
 	} else {
 		validators, err = s.Chain.GetNextBlockValidators()
 	}

--- a/pkg/core/blockchain.go
+++ b/pkg/core/blockchain.go
@@ -2697,18 +2697,20 @@ func (bc *Blockchain) GetCommittee() (keys.PublicKeys, error) {
 	return pubs, nil
 }
 
-// GetValidators returns current validators. Validators list returned from this
-// method may be updated every block (depending on register/unregister/vote
-// calls to NeoToken contract), not only once per (committee size) number of
-// blocks.
-func (bc *Blockchain) GetValidators() []*keys.PublicKey {
+// ComputeNextBlockValidators returns current validators. Validators list
+// returned from this method may be updated every block (depending on
+// register/unregister/vote calls to NeoToken contract), not only once per
+// (committee size) number of blocks.
+func (bc *Blockchain) ComputeNextBlockValidators() []*keys.PublicKey {
 	return bc.contracts.NEO.ComputeNextBlockValidators(bc.dao)
 }
 
 // GetNextBlockValidators returns next block validators. Validators list returned
 // from this method is the sorted top NumOfCNs number of public keys from the
-// current committee, thus, validators list returned from this method is being
-// updated once per (committee size) number of blocks.
+// committee of the current dBFT round (that was calculated once for the
+// CommitteeSize number of blocks), thus, validators list returned from this
+// method is being updated once per (committee size) number of blocks, but not
+// every block.
 func (bc *Blockchain) GetNextBlockValidators() ([]*keys.PublicKey, error) {
 	return bc.contracts.NEO.GetNextBlockValidatorsInternal(bc.dao), nil
 }

--- a/pkg/core/blockchain.go
+++ b/pkg/core/blockchain.go
@@ -2698,9 +2698,12 @@ func (bc *Blockchain) GetCommittee() (keys.PublicKeys, error) {
 }
 
 // ComputeNextBlockValidators returns current validators. Validators list
-// returned from this method may be updated every block (depending on
-// register/unregister/vote calls to NeoToken contract), not only once per
-// (committee size) number of blocks.
+// returned from this method is updated once per CommitteeSize number of blocks.
+// For the last block in the dBFT epoch this method returns the list of validators
+// recalculated from the latest relevant information about NEO votes; in this case
+// list of validators may differ from the one returned by GetNextBlockValidators.
+// For the not-last block of dBFT epoch this method returns the same list as
+// GetNextBlockValidators.
 func (bc *Blockchain) ComputeNextBlockValidators() []*keys.PublicKey {
 	return bc.contracts.NEO.ComputeNextBlockValidators(bc.dao)
 }

--- a/pkg/core/blockchain.go
+++ b/pkg/core/blockchain.go
@@ -2697,12 +2697,18 @@ func (bc *Blockchain) GetCommittee() (keys.PublicKeys, error) {
 	return pubs, nil
 }
 
-// GetValidators returns current validators.
-func (bc *Blockchain) GetValidators() ([]*keys.PublicKey, error) {
-	return bc.contracts.NEO.ComputeNextBlockValidators(bc.blockHeight, bc.dao)
+// GetValidators returns current validators. Validators list returned from this
+// method may be updated every block (depending on register/unregister/vote
+// calls to NeoToken contract), not only once per (committee size) number of
+// blocks.
+func (bc *Blockchain) GetValidators() []*keys.PublicKey {
+	return bc.contracts.NEO.ComputeNextBlockValidators(bc.dao)
 }
 
-// GetNextBlockValidators returns next block validators.
+// GetNextBlockValidators returns next block validators. Validators list returned
+// from this method is the sorted top NumOfCNs number of public keys from the
+// current committee, thus, validators list returned from this method is being
+// updated once per (committee size) number of blocks.
 func (bc *Blockchain) GetNextBlockValidators() ([]*keys.PublicKey, error) {
 	return bc.contracts.NEO.GetNextBlockValidatorsInternal(bc.dao), nil
 }

--- a/pkg/core/blockchain_core_test.go
+++ b/pkg/core/blockchain_core_test.go
@@ -261,7 +261,7 @@ func TestChainWithVolatileNumOfValidators(t *testing.T) {
 
 	priv0 := testchain.PrivateKeyByID(0)
 
-	vals := bc.GetValidators()
+	vals := bc.ComputeNextBlockValidators()
 	script, err := smartcontract.CreateDefaultMultiSigRedeemScript(vals)
 	require.NoError(t, err)
 	curWit := transaction.Witness{
@@ -279,7 +279,7 @@ func TestChainWithVolatileNumOfValidators(t *testing.T) {
 		}
 		// Mimic consensus.
 		if bc.config.ShouldUpdateCommitteeAt(uint32(i)) {
-			vals = bc.GetValidators()
+			vals = bc.ComputeNextBlockValidators()
 		} else {
 			vals, err = bc.GetNextBlockValidators()
 		}

--- a/pkg/core/blockchain_core_test.go
+++ b/pkg/core/blockchain_core_test.go
@@ -261,8 +261,7 @@ func TestChainWithVolatileNumOfValidators(t *testing.T) {
 
 	priv0 := testchain.PrivateKeyByID(0)
 
-	vals, err := bc.GetValidators()
-	require.NoError(t, err)
+	vals := bc.GetValidators()
 	script, err := smartcontract.CreateDefaultMultiSigRedeemScript(vals)
 	require.NoError(t, err)
 	curWit := transaction.Witness{
@@ -280,7 +279,7 @@ func TestChainWithVolatileNumOfValidators(t *testing.T) {
 		}
 		// Mimic consensus.
 		if bc.config.ShouldUpdateCommitteeAt(uint32(i)) {
-			vals, err = bc.GetValidators()
+			vals = bc.GetValidators()
 		} else {
 			vals, err = bc.GetNextBlockValidators()
 		}

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -50,7 +50,12 @@ type NeoCache struct {
 
 	votesChanged   bool
 	nextValidators keys.PublicKeys
-	validators     keys.PublicKeys
+	// validators contains cached next block validators. This list is updated every
+	// PostPersist if candidates votes ratio has been changed or register/unregister
+	// operation was performed within the persisted block. The updated value is
+	// being persisted following the standard layered DAO persist rules, so that
+	// external users will always get the proper value with upper Blockchain's DAO.
+	validators keys.PublicKeys
 	// committee contains cached committee members and their votes.
 	// It is updated once in a while depending on committee size
 	// (every 28 blocks for mainnet). It's value
@@ -269,6 +274,7 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
+		validators:      nil, // will be updated in genesis PostPersist.
 	}
 
 	// We need cache to be present in DAO before the subsequent call to `mint`.
@@ -325,6 +331,12 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache.gasPerBlock = n.getSortedGASRecordFromDAO(d)
 	cache.registerPrice = getIntWithKey(n.ID, d, []byte{prefixRegisterPrice})
 
+	numOfCNs := n.cfg.GetNumOfCNs(blockHeight + 1)
+	err := n.updateCachedValidators(d, cache, blockHeight, numOfCNs)
+	if err != nil {
+		return fmt.Errorf("failed to update next block validators cache: %w", err)
+	}
+
 	d.SetCache(n.ID, cache)
 	return nil
 }
@@ -350,6 +362,20 @@ func (n *NEO) updateCache(cache *NeoCache, cvs keysWithVotes, blockHeight uint32
 	nextVals := committee[:n.cfg.GetNumOfCNs(blockHeight+1)].Copy()
 	sort.Sort(nextVals)
 	cache.nextValidators = nextVals
+	return nil
+}
+
+// updateCachedValidators sets validators cache that will be used by external users
+// to retrieve next block validators list. Thus, it stores the list of validators
+// computed using the currently persisted block state.
+func (n *NEO) updateCachedValidators(d *dao.Simple, cache *NeoCache, blockHeight uint32, numOfCNs int) error {
+	result, _, err := n.computeCommitteeMembers(blockHeight, d)
+	if err != nil {
+		return fmt.Errorf("failed to compute committee members: %w", err)
+	}
+	result = result[:numOfCNs]
+	sort.Sort(result)
+	cache.validators = result
 	return nil
 }
 
@@ -399,6 +425,7 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 	committeeReward := new(big.Int).Mul(gas, bigCommitteeRewardRatio)
 	n.GAS.mint(ic, pubs[index].GetScriptHash(), committeeReward.Div(committeeReward, big100), false)
 
+	var isCacheRW bool
 	if n.cfg.ShouldUpdateCommitteeAt(ic.Block.Index) {
 		var voterReward = new(big.Int).Set(bigVoterRewardRatio)
 		voterReward.Mul(voterReward, gas)
@@ -408,9 +435,8 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 		voterReward.Div(voterReward, big100)
 
 		var (
-			cs        = cache.committee
-			isCacheRW bool
-			key       = make([]byte, 34)
+			cs  = cache.committee
+			key = make([]byte, 34)
 		)
 		for i := range cs {
 			if cs[i].Votes.Sign() > 0 {
@@ -437,6 +463,21 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			}
 		}
 	}
+	// Update next block validators cache for external users.
+	var (
+		h        = ic.Block.Index // consider persisting block as stored to get _next_ block validators
+		numOfCNs = n.cfg.GetNumOfCNs(h + 1)
+	)
+	if cache.validators == nil || numOfCNs != len(cache.validators) {
+		if !isCacheRW {
+			cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
+		}
+		err := n.updateCachedValidators(ic.DAO, cache, h, numOfCNs)
+		if err != nil {
+			return fmt.Errorf("failed to update next block validators cache: %w", err)
+		}
+	}
+
 	return nil
 }
 
@@ -1042,24 +1083,21 @@ func (n *NEO) getAccountState(ic *interop.Context, args []stackitem.Item) stacki
 	return item
 }
 
-// ComputeNextBlockValidators returns an actual list of current validators.
-func (n *NEO) ComputeNextBlockValidators(blockHeight uint32, d *dao.Simple) (keys.PublicKeys, error) {
-	numOfCNs := n.cfg.GetNumOfCNs(blockHeight + 1)
-	// Most of the time it should be OK with RO cache, thus try to retrieve
-	// validators without RW cache creation to avoid cached values copying.
+// ComputeNextBlockValidators computes an actual list of current validators that is
+// relevant for the latest persisted block and based on the latest changes made by
+// register/unregister/vote calls.
+// Note: this method isn't actually "computes" new committee list and calculates
+// new validators list from it. Instead, it uses cache, but the cache itself is
+// updated every block.
+func (n *NEO) ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys {
+	// It should always be OK with RO cache if using lower-layered DAO with proper
+	// cache set.
 	cache := d.GetROCache(n.ID).(*NeoCache)
-	if vals := cache.validators; vals != nil && numOfCNs == len(vals) {
-		return vals.Copy(), nil
+	if vals := cache.validators; vals != nil {
+		return vals.Copy()
 	}
-	cache = d.GetRWCache(n.ID).(*NeoCache)
-	result, _, err := n.computeCommitteeMembers(blockHeight, d)
-	if err != nil {
-		return nil, err
-	}
-	result = result[:numOfCNs]
-	sort.Sort(result)
-	cache.validators = result
-	return result, nil
+	// It's a program error not to have the right value in lower cache.
+	panic(fmt.Errorf("unexpected validators cache content: %v", cache.validators))
 }
 
 func (n *NEO) getCommittee(ic *interop.Context, _ []stackitem.Item) stackitem.Item {

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -363,7 +363,7 @@ func (n *NEO) initConfigCache(cfg config.ProtocolConfiguration) error {
 func (n *NEO) updateCache(cache *NeoCache, cvs keysWithVotes, blockHeight uint32) error {
 	cache.committee = cvs
 
-	var committee = getCommitteeMembers(cache)
+	var committee = getCommitteeMembers(cache.committee)
 	script, err := smartcontract.CreateMajorityMultiSigRedeemScript(committee.Copy())
 	if err != nil {
 		return err
@@ -431,7 +431,7 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
 func (n *NEO) PostPersist(ic *interop.Context) error {
 	gas := n.GetGASPerBlock(ic.DAO, ic.Block.Index)
 	cache := ic.DAO.GetROCache(n.ID).(*NeoCache)
-	pubs := getCommitteeMembers(cache)
+	pubs := getCommitteeMembers(cache.committee)
 	committeeSize := n.cfg.GetCommitteeSize(ic.Block.Index)
 	index := int(ic.Block.Index) % committeeSize
 	committeeReward := new(big.Int).Mul(gas, bigCommitteeRewardRatio)
@@ -1137,11 +1137,10 @@ func (n *NEO) modifyVoterTurnout(d *dao.Simple, amount *big.Int) error {
 // GetCommitteeMembers returns public keys of nodes in committee using cached value.
 func (n *NEO) GetCommitteeMembers(d *dao.Simple) keys.PublicKeys {
 	cache := d.GetROCache(n.ID).(*NeoCache)
-	return getCommitteeMembers(cache)
+	return getCommitteeMembers(cache.committee)
 }
 
-func getCommitteeMembers(cache *NeoCache) keys.PublicKeys {
-	var cvs = cache.committee
+func getCommitteeMembers(cvs keysWithVotes) keys.PublicKeys {
 	var committee = make(keys.PublicKeys, len(cvs))
 	var err error
 	for i := range committee {

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -139,9 +139,11 @@ func copyNeoCache(src, dst *NeoCache) {
 	// Can safely omit copying because the new array is created each time
 	// newEpochNextValidators list, nextValidators and committee are updated.
 	dst.nextValidators = src.nextValidators
-	dst.newEpochNextValidators = src.newEpochNextValidators
 	dst.committee = src.committee
 	dst.committeeHash = src.committeeHash
+	dst.newEpochNextValidators = src.newEpochNextValidators
+	dst.newEpochCommittee = src.newEpochCommittee
+	dst.newEpochCommitteeHash = src.newEpochCommitteeHash
 
 	dst.registerPrice = src.registerPrice
 
@@ -281,11 +283,6 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
-		// Will be updated in the last epoch block's PostPersist right before the committee update block.
-		// NeoToken's deployment (and initialization) isn't intended to be performed not-in-genesis block anyway.
-		newEpochNextValidators: nil,
-		newEpochCommittee:      nil,
-		newEpochCommitteeHash:  util.Uint160{},
 	}
 
 	// We need cache to be present in DAO before the subsequent call to `mint`.
@@ -317,6 +314,11 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	setIntWithKey(n.ID, ic.DAO, []byte{prefixRegisterPrice}, DefaultRegisterPrice)
 	cache.registerPrice = int64(DefaultRegisterPrice)
 
+	var numOfCNs = n.cfg.GetNumOfCNs(ic.Block.Index + 1)
+	err = n.updateCachedNewEpochValues(ic.DAO, cache, ic.BlockHeight(), numOfCNs)
+	if err != nil {
+		return fmt.Errorf("failed to update next block newEpoch* cache: %w", err)
+	}
 	return nil
 }
 
@@ -328,13 +330,6 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
-		// If it's a block in the middle of dBFT epoch, then no one must access
-		// these NewEpoch* fields, and they will be updated during the PostPersist of the last
-		// block in the current epoch. If it's the last block of the current epoch,
-		// then update this cache manually below.
-		newEpochNextValidators: nil,
-		newEpochCommittee:      nil,
-		newEpochCommitteeHash:  util.Uint160{},
 	}
 
 	var committee = keysWithVotes{}
@@ -349,14 +344,21 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache.gasPerBlock = n.getSortedGASRecordFromDAO(d)
 	cache.registerPrice = getIntWithKey(n.ID, d, []byte{prefixRegisterPrice})
 
-	// Update newEpoch* cache for external users if committee should be
-	// updated in the next block.
+	// Update newEpoch* cache for external users. It holds values for the previous
+	// dBFT epoch if the current one isn't yet finished.
 	if n.cfg.ShouldUpdateCommitteeAt(blockHeight + 1) {
 		var numOfCNs = n.cfg.GetNumOfCNs(blockHeight + 1)
 		err := n.updateCachedNewEpochValues(d, cache, blockHeight, numOfCNs)
 		if err != nil {
 			return fmt.Errorf("failed to update next block newEpoch* cache: %w", err)
 		}
+	} else {
+		// nextValidators, committee and committee hash are filled in by this moment
+		// via n.updateCache call.
+		cache.newEpochNextValidators = cache.nextValidators.Copy()
+		cache.newEpochCommittee = make(keysWithVotes, len(cache.committee))
+		copy(cache.newEpochCommittee, cache.committee)
+		cache.newEpochCommitteeHash = cache.committeeHash
 	}
 
 	d.SetCache(n.ID, cache)
@@ -418,15 +420,10 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
 		cache := ic.DAO.GetRWCache(n.ID).(*NeoCache)
 		// Cached newEpoch* values always have proper value set (either by PostPersist
 		// during the last epoch block handling or by initialization code).
-		oldKeys := cache.nextValidators
-		oldCom := cache.committee
-		if n.cfg.GetNumOfCNs(ic.Block.Index) != len(oldKeys) ||
-			n.cfg.GetCommitteeSize(ic.Block.Index) != len(oldCom) {
-			cache.nextValidators = cache.newEpochNextValidators
-			cache.committee = cache.newEpochCommittee
-			cache.committeeHash = cache.newEpochCommitteeHash
-			cache.votesChanged = false
-		}
+		cache.nextValidators = cache.newEpochNextValidators
+		cache.committee = cache.newEpochCommittee
+		cache.committeeHash = cache.newEpochCommitteeHash
+		cache.votesChanged = false
 
 		// We need to put in storage anyway, as it affects dumps
 		ic.DAO.PutStorageItem(n.ID, prefixCommittee, cache.committee.Bytes(ic.DAO.GetItemCtx()))
@@ -490,7 +487,9 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			h        = ic.Block.Index // consider persisting block as stored to get _next_ block newEpochNextValidators
 			numOfCNs = n.cfg.GetNumOfCNs(h + 1)
 		)
-		if cache.newEpochNextValidators == nil || numOfCNs != len(cache.newEpochNextValidators) {
+		if cache.votesChanged ||
+			numOfCNs != len(cache.newEpochNextValidators) ||
+			n.cfg.GetCommitteeSize(h+1) != len(cache.newEpochCommittee) {
 			if !isCacheRW {
 				cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
 			}
@@ -839,7 +838,9 @@ func (n *NEO) UnregisterCandidateInternal(ic *interop.Context, pub *keys.PublicK
 		return nil
 	}
 	cache := ic.DAO.GetRWCache(n.ID).(*NeoCache)
-	cache.newEpochNextValidators = nil
+	// Not only current committee/validators cache is interested in votesChanged, but also
+	// newEpoch cache, thus, modify votesChanged to update the latter.
+	cache.votesChanged = true
 	c := new(candidate).FromBytes(si)
 	emitEvent := c.Registered
 	c.Registered = false
@@ -1120,7 +1121,9 @@ func (n *NEO) ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys {
 		return vals.Copy()
 	}
 	// It's a caller's program error to call ComputeNextBlockValidators not having
-	// the right value in lower cache.
+	// the right value in lower cache. With the current scheme of handling
+	// newEpochNextValidators cache this code is expected to be unreachable, but
+	// let's have this panic here just in case.
 	panic("bug: unexpected external call to newEpochNextValidators cache")
 }
 

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -50,11 +50,13 @@ type NeoCache struct {
 
 	votesChanged   bool
 	nextValidators keys.PublicKeys
-	// validators contains cached next block validators. This list is updated every
-	// PostPersist if candidates votes ratio has been changed or register/unregister
-	// operation was performed within the persisted block. The updated value is
-	// being persisted following the standard layered DAO persist rules, so that
-	// external users will always get the proper value with upper Blockchain's DAO.
+	// validators contains cached next block validators. This list is updated once
+	// per dBFT epoch in PostPersist of the last block in the epoch if candidates
+	// votes ratio has been changed or register/unregister operation was performed
+	// within the last processed epoch. The updated value is being persisted
+	// following the standard layered DAO persist rules, so that external users
+	// will get the proper value with upper Blockchain's DAO (but this value is
+	// relevant only by the moment of first epoch block creation).
 	validators keys.PublicKeys
 	// committee contains cached committee members and their votes.
 	// It is updated once in a while depending on committee size
@@ -274,7 +276,7 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
-		validators:      nil, // will be updated in genesis PostPersist.
+		validators:      nil, // will be updated in the last epoch block's PostPersist right before the committee update block.
 	}
 
 	// We need cache to be present in DAO before the subsequent call to `mint`.
@@ -317,6 +319,11 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
+		// If it's a block in the middle of dbft epoch, then no one must access
+		// this field, and it will be updated during the PostPersist of the last
+		// block in the current epoch. If it's the laast block of the current epoch,
+		// then update this cache manually below.
+		validators: nil,
 	}
 
 	var committee = keysWithVotes{}
@@ -331,10 +338,14 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache.gasPerBlock = n.getSortedGASRecordFromDAO(d)
 	cache.registerPrice = getIntWithKey(n.ID, d, []byte{prefixRegisterPrice})
 
-	numOfCNs := n.cfg.GetNumOfCNs(blockHeight + 1)
-	err := n.updateCachedValidators(d, cache, blockHeight, numOfCNs)
-	if err != nil {
-		return fmt.Errorf("failed to update next block validators cache: %w", err)
+	// Update next block validators cache for external users if committee should be
+	// updated in the next block.
+	if n.cfg.ShouldUpdateCommitteeAt(blockHeight + 1) {
+		var numOfCNs = n.cfg.GetNumOfCNs(blockHeight + 1)
+		err := n.updateCachedValidators(d, cache, blockHeight, numOfCNs)
+		if err != nil {
+			return fmt.Errorf("failed to update next block validators cache: %w", err)
+		}
 	}
 
 	d.SetCache(n.ID, cache)
@@ -366,8 +377,9 @@ func (n *NEO) updateCache(cache *NeoCache, cvs keysWithVotes, blockHeight uint32
 }
 
 // updateCachedValidators sets validators cache that will be used by external users
-// to retrieve next block validators list. Thus, it stores the list of validators
-// computed using the currently persisted block state.
+// to retrieve next block validators list of the next dBFT epoch that wasn't yet
+// started. Thus, it stores the list of validators computed using the persisted
+// blocks state of the latest epoch.
 func (n *NEO) updateCachedValidators(d *dao.Simple, cache *NeoCache, blockHeight uint32, numOfCNs int) error {
 	result, _, err := n.computeCommitteeMembers(blockHeight, d)
 	if err != nil {
@@ -463,18 +475,21 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			}
 		}
 	}
-	// Update next block validators cache for external users.
-	var (
-		h        = ic.Block.Index // consider persisting block as stored to get _next_ block validators
-		numOfCNs = n.cfg.GetNumOfCNs(h + 1)
-	)
-	if cache.validators == nil || numOfCNs != len(cache.validators) {
-		if !isCacheRW {
-			cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
-		}
-		err := n.updateCachedValidators(ic.DAO, cache, h, numOfCNs)
-		if err != nil {
-			return fmt.Errorf("failed to update next block validators cache: %w", err)
+	// Update next block validators cache for external users if committee should be
+	// updated in the next block.
+	if n.cfg.ShouldUpdateCommitteeAt(ic.Block.Index + 1) {
+		var (
+			h        = ic.Block.Index // consider persisting block as stored to get _next_ block validators
+			numOfCNs = n.cfg.GetNumOfCNs(h + 1)
+		)
+		if cache.validators == nil || numOfCNs != len(cache.validators) {
+			if !isCacheRW {
+				cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
+			}
+			err := n.updateCachedValidators(ic.DAO, cache, h, numOfCNs)
+			if err != nil {
+				return fmt.Errorf("failed to update next block validators cache: %w", err)
+			}
 		}
 	}
 
@@ -1084,11 +1099,11 @@ func (n *NEO) getAccountState(ic *interop.Context, args []stackitem.Item) stacki
 }
 
 // ComputeNextBlockValidators computes an actual list of current validators that is
-// relevant for the latest persisted block and based on the latest changes made by
-// register/unregister/vote calls.
+// relevant for the latest processed dBFT epoch and based on the changes made by
+// register/unregister/vote calls during the latest epoch.
 // Note: this method isn't actually "computes" new committee list and calculates
-// new validators list from it. Instead, it uses cache, but the cache itself is
-// updated every block.
+// new validators list from it. Instead, it uses cache, and the cache itself is
+// updated during the PostPersist of the last block of every epoch.
 func (n *NEO) ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys {
 	// It should always be OK with RO cache if using lower-layered DAO with proper
 	// cache set.
@@ -1096,8 +1111,9 @@ func (n *NEO) ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys {
 	if vals := cache.validators; vals != nil {
 		return vals.Copy()
 	}
-	// It's a program error not to have the right value in lower cache.
-	panic(fmt.Errorf("unexpected validators cache content: %v", cache.validators))
+	// It's a caller's program error to call ComputeNextBlockValidators not having
+	// the right value in lower cache.
+	panic("bug: unexpected external call to validators cache")
 }
 
 func (n *NEO) getCommittee(ic *interop.Context, _ []stackitem.Item) stackitem.Item {

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -63,8 +63,13 @@ type NeoCache struct {
 	// (every 28 blocks for mainnet). It's value
 	// is always equal to the value stored by `prefixCommittee`.
 	committee keysWithVotes
+	// newEpochCommittee contains cached committee members updated once per dBFT
+	// epoch in PostPersist of the last block in the epoch.
+	newEpochCommittee keysWithVotes
 	// committeeHash contains the script hash of the committee.
 	committeeHash util.Uint160
+	// newEpochCommitteeHash contains the script hash of the newEpochCommittee.
+	newEpochCommitteeHash util.Uint160
 
 	// gasPerVoteCache contains the last updated value of GAS per vote reward for candidates.
 	// It is set in state-modifying methods only and read in `PostPersist`, thus is not protected
@@ -274,9 +279,13 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	}
 
 	cache := &NeoCache{
-		gasPerVoteCache:        make(map[string]big.Int),
-		votesChanged:           true,
-		newEpochNextValidators: nil, // will be updated in the last epoch block's PostPersist right before the committee update block.
+		gasPerVoteCache: make(map[string]big.Int),
+		votesChanged:    true,
+		// Will be updated in the last epoch block's PostPersist right before the committee update block.
+		// NeoToken's deployment (and initialization) isn't intended to be performed not-in-genesis block anyway.
+		newEpochNextValidators: nil,
+		newEpochCommittee:      nil,
+		newEpochCommitteeHash:  util.Uint160{},
 	}
 
 	// We need cache to be present in DAO before the subsequent call to `mint`.
@@ -319,11 +328,13 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache := &NeoCache{
 		gasPerVoteCache: make(map[string]big.Int),
 		votesChanged:    true,
-		// If it's a block in the middle of dbft epoch, then no one must access
-		// this field, and it will be updated during the PostPersist of the last
-		// block in the current epoch. If it's the laast block of the current epoch,
+		// If it's a block in the middle of dBFT epoch, then no one must access
+		// these NewEpoch* fields, and they will be updated during the PostPersist of the last
+		// block in the current epoch. If it's the last block of the current epoch,
 		// then update this cache manually below.
 		newEpochNextValidators: nil,
+		newEpochCommittee:      nil,
+		newEpochCommitteeHash:  util.Uint160{},
 	}
 
 	var committee = keysWithVotes{}
@@ -338,13 +349,13 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 	cache.gasPerBlock = n.getSortedGASRecordFromDAO(d)
 	cache.registerPrice = getIntWithKey(n.ID, d, []byte{prefixRegisterPrice})
 
-	// Update next block validators cache for external users if committee should be
+	// Update newEpoch* cache for external users if committee should be
 	// updated in the next block.
 	if n.cfg.ShouldUpdateCommitteeAt(blockHeight + 1) {
 		var numOfCNs = n.cfg.GetNumOfCNs(blockHeight + 1)
-		err := n.updateCachedValidators(d, cache, blockHeight, numOfCNs)
+		err := n.updateCachedNewEpochValues(d, cache, blockHeight, numOfCNs)
 		if err != nil {
-			return fmt.Errorf("failed to update next block newEpochNextValidators cache: %w", err)
+			return fmt.Errorf("failed to update next block newEpoch* cache: %w", err)
 		}
 	}
 
@@ -376,37 +387,28 @@ func (n *NEO) updateCache(cache *NeoCache, cvs keysWithVotes, blockHeight uint32
 	return nil
 }
 
-// updateCachedValidators sets newEpochNextValidators cache that will be used by external users
-// to retrieve next block validators list of the next dBFT epoch that wasn't yet
-// started. Thus, it stores the list of validators computed using the persisted
-// blocks state of the latest epoch.
-func (n *NEO) updateCachedValidators(d *dao.Simple, cache *NeoCache, blockHeight uint32, numOfCNs int) error {
-	result, _, err := n.computeCommitteeMembers(blockHeight, d)
+// updateCachedNewEpochValues sets newEpochNextValidators, newEpochCommittee and
+// newEpochCommitteeHash cache that will be used by external users to retrieve
+// next block validators list of the next dBFT epoch that wasn't yet started and
+// will be used by corresponding values initialisation on the next epoch start.
+// The updated new epoch cached values computed using the persisted blocks state
+// of the latest epoch.
+func (n *NEO) updateCachedNewEpochValues(d *dao.Simple, cache *NeoCache, blockHeight uint32, numOfCNs int) error {
+	committee, cvs, err := n.computeCommitteeMembers(blockHeight, d)
 	if err != nil {
 		return fmt.Errorf("failed to compute committee members: %w", err)
 	}
-	result = result[:numOfCNs]
-	sort.Sort(result)
-	cache.newEpochNextValidators = result
-	return nil
-}
+	cache.newEpochCommittee = cvs
 
-func (n *NEO) updateCommittee(cache *NeoCache, ic *interop.Context) error {
-	if !cache.votesChanged {
-		// We need to put in storage anyway, as it affects dumps
-		ic.DAO.PutStorageItem(n.ID, prefixCommittee, cache.committee.Bytes(ic.DAO.GetItemCtx()))
-		return nil
-	}
-
-	_, cvs, err := n.computeCommitteeMembers(ic.BlockHeight(), ic.DAO)
+	script, err := smartcontract.CreateMajorityMultiSigRedeemScript(committee.Copy())
 	if err != nil {
 		return err
 	}
-	if err := n.updateCache(cache, cvs, ic.BlockHeight()); err != nil {
-		return err
-	}
-	cache.votesChanged = false
-	ic.DAO.PutStorageItem(n.ID, prefixCommittee, cvs.Bytes(ic.DAO.GetItemCtx()))
+	cache.newEpochCommitteeHash = hash.Hash160(script)
+
+	nextVals := committee[:numOfCNs].Copy()
+	sort.Sort(nextVals)
+	cache.newEpochNextValidators = nextVals
 	return nil
 }
 
@@ -414,15 +416,20 @@ func (n *NEO) updateCommittee(cache *NeoCache, ic *interop.Context) error {
 func (n *NEO) OnPersist(ic *interop.Context) error {
 	if n.cfg.ShouldUpdateCommitteeAt(ic.Block.Index) {
 		cache := ic.DAO.GetRWCache(n.ID).(*NeoCache)
+		// Cached newEpoch* values always have proper value set (either by PostPersist
+		// during the last epoch block handling or by initialization code).
 		oldKeys := cache.nextValidators
 		oldCom := cache.committee
 		if n.cfg.GetNumOfCNs(ic.Block.Index) != len(oldKeys) ||
 			n.cfg.GetCommitteeSize(ic.Block.Index) != len(oldCom) {
-			cache.votesChanged = true
+			cache.nextValidators = cache.newEpochNextValidators
+			cache.committee = cache.newEpochCommittee
+			cache.committeeHash = cache.newEpochCommitteeHash
+			cache.votesChanged = false
 		}
-		if err := n.updateCommittee(cache, ic); err != nil {
-			return err
-		}
+
+		// We need to put in storage anyway, as it affects dumps
+		ic.DAO.PutStorageItem(n.ID, prefixCommittee, cache.committee.Bytes(ic.DAO.GetItemCtx()))
 	}
 	return nil
 }
@@ -475,8 +482,9 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			}
 		}
 	}
-	// Update next block newEpochNextValidators cache for external users if committee should be
-	// updated in the next block.
+	// Update newEpoch cache for external users and further committee, committeeHash
+	// and nextBlockValidators cache initialisation if committee should be updated in
+	// the next block.
 	if n.cfg.ShouldUpdateCommitteeAt(ic.Block.Index + 1) {
 		var (
 			h        = ic.Block.Index // consider persisting block as stored to get _next_ block newEpochNextValidators
@@ -486,9 +494,9 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			if !isCacheRW {
 				cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
 			}
-			err := n.updateCachedValidators(ic.DAO, cache, h, numOfCNs)
+			err := n.updateCachedNewEpochValues(ic.DAO, cache, h, numOfCNs)
 			if err != nil {
-				return fmt.Errorf("failed to update next block newEpochNextValidators cache: %w", err)
+				return fmt.Errorf("failed to update next block newEpoch* cache: %w", err)
 			}
 		}
 	}

--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -50,14 +50,14 @@ type NeoCache struct {
 
 	votesChanged   bool
 	nextValidators keys.PublicKeys
-	// validators contains cached next block validators. This list is updated once
+	// newEpochNextValidators contains cached next block newEpochNextValidators. This list is updated once
 	// per dBFT epoch in PostPersist of the last block in the epoch if candidates
 	// votes ratio has been changed or register/unregister operation was performed
 	// within the last processed epoch. The updated value is being persisted
 	// following the standard layered DAO persist rules, so that external users
 	// will get the proper value with upper Blockchain's DAO (but this value is
 	// relevant only by the moment of first epoch block creation).
-	validators keys.PublicKeys
+	newEpochNextValidators keys.PublicKeys
 	// committee contains cached committee members and their votes.
 	// It is updated once in a while depending on committee size
 	// (every 28 blocks for mainnet). It's value
@@ -132,9 +132,9 @@ func (c *NeoCache) Copy() dao.NativeContractCache {
 func copyNeoCache(src, dst *NeoCache) {
 	dst.votesChanged = src.votesChanged
 	// Can safely omit copying because the new array is created each time
-	// validators list, nextValidators and committee are updated.
+	// newEpochNextValidators list, nextValidators and committee are updated.
 	dst.nextValidators = src.nextValidators
-	dst.validators = src.validators
+	dst.newEpochNextValidators = src.newEpochNextValidators
 	dst.committee = src.committee
 	dst.committeeHash = src.committeeHash
 
@@ -274,9 +274,9 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	}
 
 	cache := &NeoCache{
-		gasPerVoteCache: make(map[string]big.Int),
-		votesChanged:    true,
-		validators:      nil, // will be updated in the last epoch block's PostPersist right before the committee update block.
+		gasPerVoteCache:        make(map[string]big.Int),
+		votesChanged:           true,
+		newEpochNextValidators: nil, // will be updated in the last epoch block's PostPersist right before the committee update block.
 	}
 
 	// We need cache to be present in DAO before the subsequent call to `mint`.
@@ -323,7 +323,7 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 		// this field, and it will be updated during the PostPersist of the last
 		// block in the current epoch. If it's the laast block of the current epoch,
 		// then update this cache manually below.
-		validators: nil,
+		newEpochNextValidators: nil,
 	}
 
 	var committee = keysWithVotes{}
@@ -344,7 +344,7 @@ func (n *NEO) InitializeCache(blockHeight uint32, d *dao.Simple) error {
 		var numOfCNs = n.cfg.GetNumOfCNs(blockHeight + 1)
 		err := n.updateCachedValidators(d, cache, blockHeight, numOfCNs)
 		if err != nil {
-			return fmt.Errorf("failed to update next block validators cache: %w", err)
+			return fmt.Errorf("failed to update next block newEpochNextValidators cache: %w", err)
 		}
 	}
 
@@ -376,7 +376,7 @@ func (n *NEO) updateCache(cache *NeoCache, cvs keysWithVotes, blockHeight uint32
 	return nil
 }
 
-// updateCachedValidators sets validators cache that will be used by external users
+// updateCachedValidators sets newEpochNextValidators cache that will be used by external users
 // to retrieve next block validators list of the next dBFT epoch that wasn't yet
 // started. Thus, it stores the list of validators computed using the persisted
 // blocks state of the latest epoch.
@@ -387,7 +387,7 @@ func (n *NEO) updateCachedValidators(d *dao.Simple, cache *NeoCache, blockHeight
 	}
 	result = result[:numOfCNs]
 	sort.Sort(result)
-	cache.validators = result
+	cache.newEpochNextValidators = result
 	return nil
 }
 
@@ -475,20 +475,20 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 			}
 		}
 	}
-	// Update next block validators cache for external users if committee should be
+	// Update next block newEpochNextValidators cache for external users if committee should be
 	// updated in the next block.
 	if n.cfg.ShouldUpdateCommitteeAt(ic.Block.Index + 1) {
 		var (
-			h        = ic.Block.Index // consider persisting block as stored to get _next_ block validators
+			h        = ic.Block.Index // consider persisting block as stored to get _next_ block newEpochNextValidators
 			numOfCNs = n.cfg.GetNumOfCNs(h + 1)
 		)
-		if cache.validators == nil || numOfCNs != len(cache.validators) {
+		if cache.newEpochNextValidators == nil || numOfCNs != len(cache.newEpochNextValidators) {
 			if !isCacheRW {
 				cache = ic.DAO.GetRWCache(n.ID).(*NeoCache)
 			}
 			err := n.updateCachedValidators(ic.DAO, cache, h, numOfCNs)
 			if err != nil {
-				return fmt.Errorf("failed to update next block validators cache: %w", err)
+				return fmt.Errorf("failed to update next block newEpochNextValidators cache: %w", err)
 			}
 		}
 	}
@@ -831,7 +831,7 @@ func (n *NEO) UnregisterCandidateInternal(ic *interop.Context, pub *keys.PublicK
 		return nil
 	}
 	cache := ic.DAO.GetRWCache(n.ID).(*NeoCache)
-	cache.validators = nil
+	cache.newEpochNextValidators = nil
 	c := new(candidate).FromBytes(si)
 	emitEvent := c.Registered
 	c.Registered = false
@@ -957,7 +957,7 @@ func (n *NEO) ModifyAccountVotes(acc *state.NEOBalance, d *dao.Simple, value *bi
 				return nil
 			}
 		}
-		cache.validators = nil
+		cache.newEpochNextValidators = nil
 		return putConvertibleToDAO(n.ID, d, key, cd)
 	}
 	return nil
@@ -1108,12 +1108,12 @@ func (n *NEO) ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys {
 	// It should always be OK with RO cache if using lower-layered DAO with proper
 	// cache set.
 	cache := d.GetROCache(n.ID).(*NeoCache)
-	if vals := cache.validators; vals != nil {
+	if vals := cache.newEpochNextValidators; vals != nil {
 		return vals.Copy()
 	}
 	// It's a caller's program error to call ComputeNextBlockValidators not having
 	// the right value in lower cache.
-	panic("bug: unexpected external call to validators cache")
+	panic("bug: unexpected external call to newEpochNextValidators cache")
 }
 
 func (n *NEO) getCommittee(ic *interop.Context, _ []stackitem.Item) stackitem.Item {

--- a/pkg/core/native/native_test/neo_test.go
+++ b/pkg/core/native/native_test/neo_test.go
@@ -138,8 +138,7 @@ func TestNEO_Vote(t *testing.T) {
 	require.NoError(t, err)
 	standBySorted = standBySorted[:validatorsCount]
 	sort.Sort(standBySorted)
-	pubs, err := e.Chain.GetValidators()
-	require.NoError(t, err)
+	pubs := e.Chain.GetValidators()
 	require.Equal(t, standBySorted, keys.PublicKeys(pubs))
 
 	// voters vote for candidates. The aim of this test is to check if voting
@@ -176,7 +175,7 @@ func TestNEO_Vote(t *testing.T) {
 	}
 
 	// We still haven't voted enough validators in.
-	pubs, err = e.Chain.GetValidators()
+	pubs = e.Chain.GetValidators()
 	require.NoError(t, err)
 	require.Equal(t, standBySorted, keys.PublicKeys(pubs))
 
@@ -267,8 +266,7 @@ func TestNEO_Vote(t *testing.T) {
 
 	advanceChain(t)
 
-	pubs, err = e.Chain.GetValidators()
-	require.NoError(t, err)
+	pubs = e.Chain.GetValidators()
 	for i := range pubs {
 		require.NotEqual(t, candidates[0], pubs[i])
 		require.NotEqual(t, candidates[len(candidates)-1], pubs[i])

--- a/pkg/core/native/native_test/neo_test.go
+++ b/pkg/core/native/native_test/neo_test.go
@@ -138,7 +138,7 @@ func TestNEO_Vote(t *testing.T) {
 	require.NoError(t, err)
 	standBySorted = standBySorted[:validatorsCount]
 	sort.Sort(standBySorted)
-	pubs := e.Chain.GetValidators()
+	pubs := e.Chain.ComputeNextBlockValidators()
 	require.Equal(t, standBySorted, keys.PublicKeys(pubs))
 
 	// voters vote for candidates. The aim of this test is to check if voting
@@ -175,7 +175,7 @@ func TestNEO_Vote(t *testing.T) {
 	}
 
 	// We still haven't voted enough validators in.
-	pubs = e.Chain.GetValidators()
+	pubs = e.Chain.ComputeNextBlockValidators()
 	require.NoError(t, err)
 	require.Equal(t, standBySorted, keys.PublicKeys(pubs))
 
@@ -266,7 +266,7 @@ func TestNEO_Vote(t *testing.T) {
 
 	advanceChain(t)
 
-	pubs = e.Chain.GetValidators()
+	pubs = e.Chain.ComputeNextBlockValidators()
 	for i := range pubs {
 		require.NotEqual(t, candidates[0], pubs[i])
 		require.NotEqual(t, candidates[len(candidates)-1], pubs[i])

